tests/test_make.py

```diff
@@ -2131,3 +2138,58 @@ def __setstate__(self, state):
 
         assert True is i.called
         assert None is getattr(C(), "__getstate__", None)
+
+
+class TestTransformHook:
+    """Tests for `attrs(tranform_hook=func)`"""
+
+    def test_hook_applied(self):
+        """The transform hook must be applied to all attributes.  Types will
+        not be available here (they only work with type annotations)."""
+        results = []
+
+        def hook(name, ca, type):
+            results.append((name, type, isinstance(ca, _CountingAttr)))
+
+        @attr.s(transform_hook=hook)
+        class C:
+            x = attr.ib()
+            y = attr.ib(type=int)
+
+        assert results == [
+            ('x', None, True),
+            ('y', None, True),
+        ]
+
+    @pytest.mark.skipif(PY2, reason="Auto-detection is Python 3-only.")
+    def test_hook_applied_auto_attrib(self):
+        """The transform hook must be applied to all attributes and type
+        annotations must be detected."""
+        results = []
+
+        def hook(name, ca, type):
+            results.append((name, type, isinstance(ca, _CountingAttr)))
+
+        @attr.s(auto_attribs=True, transform_hook=hook)
+        class C:
+            x: int
+            y: str = attr.ib()
+
+        assert results == [
+            ('x', int, True),
+            ('y', str, True),
+        ]
+
+    @pytest.mark.skipif(PY2, reason="Auto-detection is Python 3-only.")
+    def test_hook_applied_modify_attrib(self):
+        """The transform hook must be able to modify attributes."""
+        def hook(name, ca, type):
+            ca.converter = type
```

Comment:
Hrm… if I were writing a hook like this, I'd be type hinting everything, and the type of `ca` here is `_CountingAttr`, which is private.  That seems like a problem.

IIRC we've avoided making `_CountingAttr` public, so this API would require us to at least define an interface, no?

Reply:
Yes, that’s true. I also stumbled upon this when I was writing the docstring. Making an Interface seems like a reasonable idea. We wouldn’t to expose the actual `_CountingAttr` than but mypy would be happy nonetheless.